qi-doc/scribblings/principles.scrbl

```diff
@@ -55,6 +67,107 @@ Consider this example:
+
+In the case where we use the @racket[effect] form on its own as in @racket[(effect displayln)], the implicit associated function is the identity flow, @racket[_].
+
+@subsection{Upstream and Downstream}
+
+We already saw how a flow can be thought of as a @seclink["Flows_as_Graphs"]{directed graph}. This naturally suggests that some flows are upstream (or downstream) of others in terms of this directionality. Let's define this relation more precisely, as it will be useful to us.
+
+@definition["Upstream and downstream"]{A flow invocation @${f} is @deftech{upstream} of another invocation @${g} if the output of @${f} is @emph{necessary} to determining the input of @${g}. Conversely, @${g} is @deftech{downstream} of @${f}. We will denote this relation @${f < g}.}
+
+Note that this definition relates flow @emph{invocations} rather than @tech{flows} themselves. For now, we need not worry about this distinction, but we will soon see why it matters.
+
+In terms of the @tech{paths} that could be traced over a flow, the ordering implied by @racket[~>] naturally shows us many members of this relation: flows that come later in the sequence in a @racket[~>] form are downstream of those that come earlier, because the output of earlier flows is needed to determine the input to later flows.
+
+In the above example, @racket[filter] and @racket[map] are obviously ordered by @racket[~>] in this way, so that @racket[(filter my-odd?)] is upstream of @racket[(map my-sqr)]. But it's not so obvious how @racket[my-odd?] and @racket[my-sqr] should be treated. These are employed "internally" by the higher-order flows @racket[filter] and @racket[map], and are not directly ordered by the @racket[~>] form. Should @racket[my-odd?] be considered to be upstream of @racket[my-sqr] here?
+
+This is where the distinction between flows and flow invocations comes into play. In fact, not all invocations of @racket[my-odd?] are upstream of any particular invocation of @racket[my-sqr]. Rather, specific invocations of @racket[my-sqr] that use values computed by individual invocations of @racket[my-odd?] are downstream of those invocations, and notably, these invocations involve the individual elements of the input list rather than the entire list, so that the computational dependency expressed by this relation is as fine-grained as possible.
+
+For instance, for an input list @racket[(list 1 2 3)], @racket[(my-odd? 1)] is @tech{upstream} of @racket[(my-sqr 1)], and likewise, @racket[(my-odd? 3)] is @tech{upstream} of @racket[(my-sqr 3)].
+
+That brings us to the guarantee that Qi provides in this case (and in general).
+
+@subsection{Qi's Guarantee on Effects}
+
+@definition["Effect locality"]{For @tech{flow} invocations @${f} and @${g} and corresponding effects @${ε(f)} and @${ε(g)},
+
+@$${f < g ⇒ ε(f) < ε(g)}
+
+where @${<} on the left denotes the relation of being upstream, and @${<} on the right denotes one effect happening before another.
```

Comment:
Punning these symbols is convenient, but it can be bothersome (it requires the reader to work out types to know which symbol is referencing which operation). How about a subscript or a different operator for each relation?

Reply:
I can imagine it could be confusing (but I think I'd need to do actual proofs or something to see if one is easier than the other!), but for the moment I feel punning on the familiar order symbol is easier on the reader since we don't actually use the RHS `<` anywhere else besides here, and subscripts could make it more technical than we currently have a use for.